internal/dataplane/client/client.go

```diff
@@ -2,64 +2,212 @@ package client
 
 import (
 	"context"
+	"errors"
 	"fmt"
+	"sync"
 
+	"github.com/go-logr/logr"
 	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
 	corev1 "k8s.io/api/core/v1"
-	"sigs.k8s.io/controller-runtime/pkg/client"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"sigs.k8s.io/controller-runtime/pkg/log"
 
 	"github.com/kong/blixt/pkg/vars"
 )
 
-// NewDataPlaneClient provides a new client for communicating with the grpc API
-// of the data-plane given a function which can provide the API endpoint.
-func NewDataPlaneClient(ctx context.Context, c client.Client) (BackendsClient, error) {
-	endpoints, err := GetDataPlaneEndpointsFromDefaultPods(ctx, c)
+// clientInfo encapsulates the gathered information about a BackendsClient
+// along with the gRPC client connection.
+type clientInfo struct {
+	conn   *grpc.ClientConn
+	client BackendsClient
+	name   string
+}
+
+// BackendsClientManager is managing the connections and interactions with
+// the available BackendsClient servers.
+type BackendsClientManager struct {
+	log       logr.Logger
+	clientset *kubernetes.Clientset
+
+	mu      sync.RWMutex
+	clients map[types.NamespacedName]clientInfo
+}
+
+// NewBackendsClientManager returns an initialized instance of BackendsClientManager.
+func NewBackendsClientManager(config *rest.Config) (*BackendsClientManager, error) {
+	clientset, err := kubernetes.NewForConfig(config)
 	if err != nil {
 		return nil, err
 	}
 
-	if len(endpoints) < 1 {
-		return nil, fmt.Errorf("no endpoints could be found for the dataplane API")
+	return &BackendsClientManager{
+		log:       log.FromContext(context.Background()),
+		clientset: clientset,
+		mu:        sync.RWMutex{},
+		clients:   map[types.NamespacedName]clientInfo{},
+	}, nil
+}
+
+func (c *BackendsClientManager) SetClientsList(ctx context.Context, readyPods map[types.NamespacedName]corev1.Pod) (bool, error) {
+	// TODO: close and connect to the different clients concurrently.
```

Comment:
We need to resolve this TODO in one of two ways:

1. implement the solution now, as a part of this PR
2. if we can live without the solution temporarily, create a follow-up issue that explains what is needed, and why, and what the downsides are of not having it, and link it here in the comment so that we don't lose track of this need.

Reply:
I've opened https://github.com/kubernetes-sigs/blixt/issues/127 to address the issue